### PR TITLE
Improve function least/greatest for nullable numberic type arguments 

### PR DESCRIPTION
**Pull request by KevinyhZou**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Improve function least/greatest for nullable numberic type arguments

If the arguments is numberic but nullable, and has 2 arguments, it would use `FunctionLeastGreatestGeneric` to do the comparsion, which is not efficient. and use `GreatestBaseImpl/LeastBaseImpl` for `FunctionBinaryArithmetic` would have better performance. So here we improve these two function by allow this.

and compare the performance below for greatest function, sql `select count(1) from test where greatest(x1,x2)>1` and the fields `x1`, `x2` is `nullable(Int32)` and `nullable(float32)` type, the test table has 1000000 rows.

before the pr
```
1 row in set. Elapsed: 0.271 sec. Processed 10.00 million rows, 100.00 MB (36.90 million rows/s., 368.99 MB/s.)
1 row in set. Elapsed: 0.244 sec. Processed 10.00 million rows, 100.00 MB (41.06 million rows/s., 410.57 MB/s.)
1 row in set. Elapsed: 0.243 sec. Processed 10.00 million rows, 100.00 MB (41.16 million rows/s., 411.59 MB/s.)
```

and after this pr
```
1 row in set. Elapsed: 0.130 sec. Processed 10.00 million rows, 100.00 MB (76.97 million rows/s., 769.66 MB/s.)
1 row in set. Elapsed: 0.133 sec. Processed 10.00 million rows, 100.00 MB (75.26 million rows/s., 752.62 MB/s.)
1 row in set. Elapsed: 0.125 sec. Processed 10.00 million rows, 100.00 MB (79.99 million rows/s., 799.86 MB/s.)
```

### Documentation entry for user-facing changes
- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

<details>
    <summary>CI Settings</summary>

**NOTE:** If your merge the PR with modified CI you **MUST KNOW** what you are doing
**NOTE:** Checked options will be applied if set before CI RunConfig/PrepareRunConfig step
- [ ]  Allow: Integration Tests
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Unit tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All with aarch64
- [ ]  Allow: All with ASAN
- [ ]  Allow: All with TSAN
- [ ]  Allow: All with Analyzer
- [ ]  Allow: All with Azure
- [ ]  Allow: Add your option here
---
- [ ]  Exclude: Fast test
- [ ]  Exclude: Integration Tests
- [ ]  Exclude: Stateless tests
- [ ]  Exclude: Stateful tests
- [ ]  Exclude: Performance tests
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN
- [ ]  Exclude: All with MSAN
- [ ]  Exclude: All with UBSAN
- [ ]  Exclude: All with Coverage
- [ ]  Exclude: All with Aarch64
---
- [ ]  do not test (only style check)
- [ ]  upload all binary artifacts from build jobs
- [ ]  disable merge-commit (no merge from master before tests)
- [ ]  disable CI cache (job reuse)
- [ ]  allow: batch 1, 2 for multi-batch jobs
- [ ]  allow: batch 3, 4
- [ ]  allow: batch 5, 6
</details>

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
*This is an automated comment for commit b79cbe955be9eb51e2f7d007524e095f046ba9c8 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/64668/b79cbe955be9eb51e2f7d007524e095f046ba9c8/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>A Sync</td><td>If it fails, ask a maintainer for help</td><td>⏳ pending</td></tr>
<tbody>
</table>
<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: src/Functions/LeastGreatestGeneric.h [hunk 1]
```diff
@@ -111,7 +111,7 @@ class LeastGreatestOverloadResolver : public IFunctionOverloadResolver
             argument_types.push_back(argument.type);
 
         /// More efficient specialization for two numeric arguments.
-        if (arguments.size() == 2 && isNumber(arguments[0].type) && isNumber(arguments[1].type))
+        if (arguments.size() == 2 && isNumber(removeNullable(arguments[0].type)) && isNumber(removeNullable(arguments[1].type)))
```
Comment (Algunenano) [inline src/Functions/LeastGreatestGeneric.h:114]: There is another condition like this in `getReturnTypeImpl` which I guess needs to be adjusted to match

**Review by Algunenano — COMMENTED** [review]
Q: Why do we need the generic comparison? What does it do differently from the binary least/greater functions?

**Comment by KevinyhZou** [comment]
The the binary least/greater is used to compare only 2 elements,  but the generic comparison can be used to compare more than 2 elements, and it seems to be more complexed than the `binary`: more `for` loop, more function call.

File: src/Functions/LeastGreatestGeneric.h (hunk 1 above)
Comment (KevinyhZou) [reply src/Functions/LeastGreatestGeneric.h:114]: done

**Comment by divanik** [comment]
Hi. LGTM, but we need to get CI checks completed. Could you merge master, please? If this doesn't help I will report this to our chat about CI bugs.

**Review by divanik — COMMENTED** [review]
Hi. LGTM, but we need to get CI checks completed. Could you pull master, please? If this doesn't help I will report this to our chat about CI bugs.

**Comment by KevinyhZou** [comment]
OK

**Comment by Algunenano** [comment]
> The the binary least/greater is used to compare only 2 elements, but the generic comparison can be used to compare more than 2 elements, to return the greatest/least element, and it seems to be more complexed than the binary: more for loop, more function call.

2 questions then:
* Why don't we always use the binary operator when we have 2 elements?
* Why don't we implement the 3+ argument functions on top of the optimized binary operator?

**Comment by KevinyhZou** [comment]
1. The binary operator is based on the function `FunctionBinaryArithmetic`, which not allow the arguments to be not numberic for `greatest`/`least`; If we want do that, then we need  modify the logic of `FunctionBinaryArithmetic`
2. If we implement the function on top of the binary operator, then need a loop to compare every 2 elements one by one, final to get the greatest/least one, this is what it do in the `FunctionLeastGreatestGeneric`, it seems not necessary。

@Algunenano

**Comment by KevinyhZou** [comment]
cc @divanik

**Comment by den-crane** [comment]
BTW 

* https://github.com/ClickHouse/ClickHouse/issues/65039

**Pull request merged**